OSC-10 Create Decision Tree component

kael89 opened the pull request:
### Issue #:
OS-10

Note: this PR is against `dev` since everything is new functionality inside `ui-components`. This means it can be safely merged without testing if you want to.

### Changes:
I created a reusable **decision tree** component. This will be used for visualising entity hierarchies in the admin panel for now, but it should be reusable for other types of hierarchies too (e.g. dashboard items, map overlays).

* Each hierarchy layer renders in a column
* The top level (root) column specifies the hierarchy we are in
* Every other child column renders recursively
* If data changes in an ancestor column (e.g. a new hierarchy is selected), all expanded descendant columns are unmounted
* The component supporters filtering in each column

The data fetching mechanism is through a `fetchData` function that is passed as a prop, and shared between the components.

```ts
interface Node {
  id: string;
  name: string;
}

type Fetcher = (
  rootNode: Node,
  node: Node,
) => {
  id: string;
  name: string;
  children: Node[];
}[];

```

```js
const fetchData = async (rootNode, node) => getEntityDescendants(rootNode?.code, node?.code);

<DecisionTree fetchData={fetchData} />
```

See storybook for usage examples of the component 

### Screenshots:
I attached a demo video to the ticket

edmofro commented:
> Note: this PR is against dev since everything is new functionality inside ui-components. This means it can be safely merged without testing if you want to.

Love this!!

kael89 commented on packages/ui-components/src/components/DecisionTree/index.js:
```diff
@@ -0,0 +1,6 @@
+ * Copyright (c) 2017 - 2023 Beyond Essential Systems Pty Ltd
+ */
+
+export * from './DecisionTree';
```
Another name I considered is `HorizontalTree`

rohan-bes replied:
I think I actually prefer `HorizontalTree`. `DecisionTree` feels as though it implies some kind of meaning about the data within the tree (ie. it's used for making decisions, eg. 'if I perform action A, event B will happen')

kael89 replied:
Happy to rename it!

kael89 commented on packages/ui-components/src/components/DecisionTree/Column.js:
```diff
@@ -0,0 +1,143 @@
+import { ColumnFilter, useFilter } from './ColumnFilter';
+import { FetchLoader } from '../FetchLoader';
+
+const CLASS_NAME = 'TupaiaDecisionTreeColumn';
```
Inspired by material-ui classes

kael89 commented on packages/ui-components/src/components/DecisionTree/Column.js:
```diff
@@ -0,0 +1,143 @@
+
+  useEffect(() => {
+    setSelected(undefined);
+  }, [data]);
```
This means that we will the selected item when `data` changes , e.g. when a different entity is selected in an ancestor column

kael89 commented on packages/ui-components/stories/decisionTree.stories.js:
```diff
@@ -0,0 +1,110 @@
+/**
```
I loved using storybook, it makes iterative component development soooo much better - and encourages you to create reusable components!!

kael89 replied:
The only thing I am not sure of is whether you want to keep all you reusable components under that same package, or whether you want to break it into smaller packages eventually

kael89 commented on packages/ui-components/src/components/DecisionTree/Column.js:
```diff
@@ -0,0 +1,143 @@
+            <List>
+              {filteredData?.length === 0 && (
+                <ListItem>
+                  <ListItemText primary="No data found" color="primary" />
```
I chose this empty state message but should confirm with the designer

kael89 commented on packages/ui-components/src/components/DecisionTree/ColumnFilter.js:
```diff
@@ -0,0 +1,58 @@
+  }
+`;
+
+export const useFilter = () => {
```
The filter is case insensitive

kael89 commented on packages/ui-components/src/components/DecisionTree/DecisionTree.js:
```diff
@@ -0,0 +1,102 @@
+            setSelectedRoot(rootNode);
+            fetchChildren(rootNode, rootNode);
+          }}
+          showExpandIcon={false}
```
Because the root component doesn't show expand icons, when we click on any of its items (hierarchies) we both select and expand it

kael89 commented on packages/ui-components/src/hooks/useData.js:
```diff
@@ -0,0 +1,37 @@
+
+import { useCallback, useState } from 'react';
+
+export const useData = fetcher => {
```
This hooks is similar to the ones in react query, if built right this can be reused in many places

kael89 commented on packages/ui-components/stories/story-utils/api/data.js:
```diff
@@ -0,0 +1,153 @@
+  { id: id(), code: 'AU_Western_Australia', name: 'Western Australia' },
+
+  // Austria
+  { id: id(), code: 'AT', name: 'Austria' },
```
I thought a bit about using `faker` for all this data, but I realised that it returns random cities for a selected country.

Seeing Athens inside Austria (this time - could be Barcelona next time) would be confusing. We wouldn't be able to tell if this works or not :smile: 

This also tests the column overflow logic as the list of European countries is long

rohan-bes reviewed (approved):
Looks great! Thanks so much for this @kael89 🎉

kael89 commented:
@rohan-bes I renamed the component to `HorizontalTree` and implemented some changes on the selection based on an updated version of the mockups.

I will merge this in `dev` as soon as you re-approve, since it doesn't have any breaking changes

rohan-bes reviewed (approved):
Looks great, thanks for the name change @kael89
